### PR TITLE
[MXNET-954] Implementation of Structured-Self-Attentive-Sentence-Embedding

### PR DESCRIPTION
**Pull request by kenjewu**
## Description ##
This is an implementation of the paper [A Structured Self-Attentive Sentence Embedding](https://arxiv.org/abs/1703.03130)，using Mxnet/Gluon.Finally, the experiment was carried out on the user review star classification task mentioned in the original paper, and used the same data set: [The reviews of Yelp Data](https://www.kaggle.com/yelp-dataset/yelp-dataset#yelp_academic_dataset_review.json). 

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
Add example: Implementation of Structured-Self-Attentive-Sentence-Embedding

### CODE REVIEW

**Comment by kalyc** [comment]
Thanks for your contribution @kenjewu 
Could you update the PR title and description to follow the template of the repository?

@mxnet-label-bot[pr-awaiting-response]

**Comment by gigasquid** [comment]
Thanks for adding the great example. The multi-language READMEs are great too 💯

**Comment by vandanavk** [comment]
@sandeep-krishnamurthy @stu1130 for review

**Comment by leezu** [comment]
Thanks @kenjewu . Would it make sense to add this to GluonNLP instead? Gluon NLP is currently developed separately from MXNet as we have faster release cycles until a stable version is reached. All scripts and examples in GluonNLP project are checked by CI so we make sure no breakage is introduced if we release new GluonNLP versions. If this is added to MXNet repo, no such checks will be made and there will be no way to address any required changes given the more infrequent MXNet release schedule.

File: example/self_attentive_sentence_embedding/README.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+This is an implementation of the paper [A Structured Self-Attentive Sentence Embedding](https://arxiv.org/abs/1703.03130). This program implements most of the details in the paper. Finally, the user reviews the emotional star ratings in the three experiments mentioned in the original paper, and used the same data set: [The reviews of Yelp Data](https://www.kaggle.com/yelp-dataset/yelp-dataset#yelp_academic_dataset_review.json). The model structure is as follows：
+
+![Bi_LSTM_Attention](./images/Bi_LSTM_Attention.png)
+
+
+
+## Requirments
+
+1. [Mxnet](https://mxnet.apache.org/)
```
Comment (eric-haibin-lin) [inline example/self_attentive_sentence_embedding/README.md]: nit pick: `Mxnet` -> `MXNet`

File: example/self_attentive_sentence_embedding/README.md [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+                           help='dropout applied to layers (0 = no dropout)')
+       parser.add_argument('--clip', type=float, default=0.5,
+                           help='clip to prevent the too large grad in LSTM')
+       parser.add_argument('--nfc', type=int, default=512,
+                           help='hidden (fully connected) layer size for classifier MLP')
+       parser.add_argument('--lr', type=float, default=.001,
+                           help='initial learning rate')
+       parser.add_argument('--epochs', type=int, default=10,
+                           help='upper epoch limit')
+       parser.add_argument('--loss-name', type=str, default='sce', help='loss function name')
```
Comment (eric-haibin-lin) [inline example/self_attentive_sentence_embedding/README.md]: Suggest to mention what loss options this script supports

File: example/self_attentive_sentence_embedding/README.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+                           help='batch size for training')
+       parser.add_argument('--class-number', type=int, default=5,
+                           help='number of classes')
+       parser.add_argument('--optimizer', type=str, default='Adam',
+                           help='type of optimizer')
+       parser.add_argument('--penalization-coeff', type=float, default=0.1,
+                           help='the penalization coefficient')
+   
+       parser.add_argument('--save', type=str, default='../models', help='path to save the final model')
+       parser.add_argument('--wv-name', type=str, choices={'glove', 'w2v', 'fasttext', 'random'},
```
Comment (eric-haibin-lin) [inline example/self_attentive_sentence_embedding/README.md]: nit pick: why is it called `wv-name`?

File: example/self_attentive_sentence_embedding/README.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+2. **Training details**
+
+   The original paper uses 500K data as the training set, 2000 data as the validation set, and 2000 as the test set. Due to personal machine restrictions, 200 K is randomly selected as the training set and 2000 data is used as the validation set in the case of ensuring the data distribution and the original data. The weight of the WeightedSoftmaxCrossEntropy is set according to the proportion of the data category. If the data is different and needs to be used To use this loss function, you need to modify the value of the set class_weight yourself.
+
+   Training usage (parameters can be customized):  
+
+   ```python
+   python train_model.py --nlayers 1 --epochs 5 --attention-hops 2 --loss-name sce
+   ```
+
```
Comment (eric-haibin-lin) [inline example/self_attentive_sentence_embedding/README.md]: What's the model accuracy if trained using this script?

**Comment by sandeep-krishnamurthy** [comment]
@kenjewu - ping :-)

File: example/self_attentive_sentence_embedding/code/prepare_data.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+            formated_data = pickle.load(f)
+    else:
+        with open(data_json_path, 'r', encoding='utf-8') as fr:
+            data = json.load(fr)
+        sentences, labels = data['texts'], data['labels']
+
+        my_vocab = get_vocab(sentences, wv_name)
+        pad_num_value = my_vocab.to_indices(PAD)
+
+        # 将输入数据转为整数索引
```
Comment (Roshrini) [inline example/self_attentive_sentence_embedding/code/prepare_data.py]: Can you please add code comments in english or have in both languages if you prefer?

File: example/self_attentive_sentence_embedding/README.md [hunk 6]
```diff
@@ -0,0 +1,107 @@
+## Implementation of Structured-Self-Attentive-Sentence-Embedding
+
+This is an implementation of the paper [A Structured Self-Attentive Sentence Embedding](https://arxiv.org/abs/1703.03130). This program implements most of the details in the paper. Finally, the user reviews the emotional star ratings in the three experiments mentioned in the original paper, and used the same data set: [The reviews of Yelp Data](https://www.kaggle.com/yelp-dataset/yelp-dataset#yelp_academic_dataset_review.json). The model structure is as follows：
+
+![Bi_LSTM_Attention](./images/Bi_LSTM_Attention.png)
+
+
+
+## Requirments
```
Comment (Roshrini) [inline example/self_attentive_sentence_embedding/README.md]: typo: Requirements

File: example/self_attentive_sentence_embedding/code/prepare_data.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+    Fill the sequence to the specified length, long truncation
+    Args:
+        sequences: A list of all sentences, a list of list
+        max_len: Specified maximum length
+        pad_value: Specified fill value
+    Returns:
+        pades_seqs: A numpy array
+    '''
+
+    # max_len = max(map(lambda x: len(x), sequences))
```
Comment (Roshrini) [inline example/self_attentive_sentence_embedding/code/prepare_data.py]: remove commented line

**Comment by Roshrini** [comment]
@sandeep-krishnamurthy Can you please change label to "pr-awaiting-response"?

**Comment by kenjewu** [comment]
This example has been modified to contribute to gluonnlp, maybe this PR can be closed.

**Comment by Roshrini** [comment]
@kenjewu Thanks for your response. Can you close this PR?

**Comment by szha** [comment]
@kenjewu thanks for the contribution!

**Pull request closed without merging**